pkg/reconcilermanager/controllers/rsync_status_watch_controller.go

```diff
@@ -0,0 +1,289 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	"kpt.dev/configsync/pkg/api/configmanagement"
+	"kpt.dev/configsync/pkg/api/configsync"
+	"kpt.dev/configsync/pkg/api/configsync/v1beta1"
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/declared"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/go-logr/logr"
+)
+
+const (
+	logFieldController    = "controller"
+	logFieldSyncName      = "sync.name"
+	logFieldSyncNamespace = "sync.namespace"
+	logFieldSyncStatus    = "status"
+	logFieldCommit        = "commit"
+
+	syncStatusFailed = "SyncFailed"
+)
+
+// RsyncStatusWatchController watches the status of RepoSync and RootSync resources.
+type RsyncStatusWatchController struct {
+	client        client.Client
+	logger        *LoggingController
+	statusTracker *StatusTracker
+	syncID        core.ID
+}
+
+// NewRsyncStatusWatchController initializes a new RsyncStatusWatchController.
+func NewRsyncStatusWatchController(mgr manager.Manager, log logr.Logger, id core.ID) *RsyncStatusWatchController {
+	return &RsyncStatusWatchController{
+		client:        mgr.GetClient(),
+		logger:        NewLoggingController(log),
+		statusTracker: NewStatusTracker(),
+		syncID:        id,
+	}
+}
+
+// newKeySelectorPredicate creates a predicate to filter events based on the object key.
+func newKeySelectorPredicate(key client.ObjectKey) predicate.Predicate {
+	return predicate.NewPredicateFuncs(func(obj client.Object) bool {
+		return client.ObjectKeyFromObject(obj) == key
+	})
+}
+
+// Register the RsyncStatusWatchController with the ReconcilerManager.
+func (c *RsyncStatusWatchController) Register(mgr controllerruntime.Manager) error {
+	cm := controllerruntime.NewControllerManagedBy(mgr).
+		Named("RsyncStatusWatchController").
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: 1,
+		})
+
+	switch c.syncID.Kind {
+	case configsync.RootSyncKind:
+		cm = cm.For(&v1beta1.RootSync{},
+			builder.WithPredicates(newKeySelectorPredicate(c.syncID.ObjectKey)))
+	case configsync.RepoSyncKind:
+		cm = cm.For(&v1beta1.RepoSync{},
+			builder.WithPredicates(newKeySelectorPredicate(c.syncID.ObjectKey)))
+	default:
+		return fmt.Errorf("invalid sync ID: %v", c.syncID)
+	}
+
+	// Watch for changes in RootSync resources
+	cm = cm.Watches(&v1beta1.RootSync{}, &handler.EnqueueRequestForObject{})
```

Comment:
I think the For calls make these Watch calls redundant.

Reply:
Done